### PR TITLE
Ukkonen dp

### PR DESCRIPTION
**Pull request by Vishal646-code**
- Added Ukkonen cutoff DP implementation for fuzzy string matching.
- Added the corresponding Ukkonen header.

### CODE REVIEW

File: libpz/fuzzy/ukkonen.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Ukkonen Cutoff DP
+#include "ukkonen.hpp"
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <cstdlib>
+#include <stdexcept>
```
Comment (AviShahCode) [inline libpz/fuzzy/ukkonen.cpp]: Use `poozle/libpz/include/pz_cxx_std.hpp`

File: libpz/fuzzy/ukkonen.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+
+bool Ukkonen::edit_within_k(const std::string& s, const std::string& t, int k) {
+    if (k < 0){
+        throw std::invalid_argument("Edit distance threshold must be non-negative");
```
Comment (AviShahCode) [inline libpz/fuzzy/ukkonen.cpp]: Add new error types in `libpz/include/pz_error.hpp` and use `report_error`

File: libpz/fuzzy/ukkonen.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+std::vector<std::pair<int, int>> Ukkonen::locate(const std::string &pattern,
+                                                const std::string &text,
+                                                int k) {
+    std::vector<std::pair<int, int>> locations;
+
+    //A negative edit distance threshold is invalid, so we throw an exception.
+    if (k < 0) {
+        throw std::invalid_argument("Edit distance threshold must be non-negative");
```
Comment (AviShahCode) [inline libpz/fuzzy/ukkonen.cpp]: Use `unsigned int` and remove checks

**Comment by AviShahCode** [comment]
1. Add more complex tests, keeping the product `nm` in order of $10^8$, and include potential edge cases.
2. For large inputs, run tests directly from text files.
3. Add basic tests like this `libpz/regex/tests/test_0_match.cpp`

File: libpz/fuzzy/output_ukkonen.txt
Comment (AviShahCode) [inline libpz/fuzzy/output_ukkonen.txt]: Move test related files to `fuzzy/tests/*` and include several testing cpp files, based on type. Example: test_0_small, test_1_large, test_2_edgecase etc.

Comment (AviShahCode) [reply libpz/fuzzy/output_ukkonen.txt]: You can refer to `regex/tests/*` as well.
